src/librustc_typeck/check/mod.rs

```diff
@@ -3819,6 +3819,69 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         pointing_at_return_type
     }
 
+    /// When encountering an fn-like ctor that needs to unify with a value, check whether calling
+    /// the ctor would successfully solve the type mismatch and if so, suggest it:
+    /// ```
+    /// fn foo(x: usize) -> usize { x }
+    /// let x: usize = foo;  // suggest calling the `foo` function: `foo(42)`
+    /// ```
+    fn suggest_fn_call(
+        &self,
+        err: &mut DiagnosticBuilder<'tcx>,
+        expr: &hir::Expr,
+        expected: Ty<'tcx>,
+        found: Ty<'tcx>,
+    ) -> bool {
+        match found.sty {
+            ty::FnDef(..) | ty::FnPtr(_) => {}
+            _ => return false,
+        }
+
+        let sig = found.fn_sig(self.tcx);
+        let sig = self
+            .replace_bound_vars_with_fresh_vars(expr.span, infer::FnCall, &sig)
+            .0;
+        let sig = self.normalize_associated_types_in(expr.span, &sig);
+        if let Ok(_) = self.try_coerce(expr, sig.output(), expected, AllowTwoPhase::No) {
+            let (mut sugg_call, applicability) = if sig.inputs().is_empty() {
+                (String::new(), Applicability::MachineApplicable)
+            } else {
+                ("...".to_owned(), Applicability::HasPlaceholders)
+            };
+            let mut msg = "call this function";
+            if let ty::FnDef(def_id, ..) = found.sty {
+                match self.tcx.hir().get_if_local(def_id) {
+                    Some(Node::Item(hir::Item {
+                        node: ItemKind::Fn(.., body_id),
+                        ..
+                    })) => {
+                        let body = self.tcx.hir().body(*body_id);
+                        sugg_call = body.arguments.iter()
+                            .map(|arg| hir::print::to_string(
+                                hir::print::NO_ANN,
+                                |s| s.print_pat(&arg.pat),
```

Comment:
I was originally going to match only on idents here and otherwise use `_`, but I thought it'd be ok to let it be this way:

```
error[E0308]: mismatched types
 --> file7.rs:8:20
  |
1 | fn foo(S(a, b): S) -> usize {
  | --------------------------- fn(S) -> usize {foo} defined here
...
8 |     let _: usize = foo;
  |                    ^^^
  |                    |
  |                    expected usize, found fn item
  |                    help: use parentheses to call this function: `foo(S(a, b))`
  |
  = note: expected type `usize`
             found type `fn(S) -> usize {foo}`
```

Reply:
Yeah it is OK but it desires a test... ;)